src/common/nvlist_utils.c

```diff
@@ -0,0 +1,389 @@
+#include "../truenas_pylibzfs.h"
+
+/*
+ * This function converts the nvlist containing user props
+ * into a python dictionary of form:
+ * {"<nvpair_name>": "<nvpair_value_string"}
+ */
+PyObject *user_props_nvlist_to_py_dict(nvlist_t *userprops)
+{
+	PyObject *d = NULL;
+	nvpair_t *elem;
+
+	d = PyDict_New();
+	if (d == NULL)
+		return NULL;
+
+	for (elem = nvlist_next_nvpair(userprops, NULL);
+	    elem != NULL;
+	    elem = nvlist_next_nvpair(userprops, elem)) {
+		const char *name = nvpair_name(elem);
+		const char *cval;
+		nvlist_t *nvl;
+
+		PYZFS_ASSERT(
+			(nvpair_type(elem) == DATA_TYPE_NVLIST),
+			"Unexpected nvpair data type in user props"
+
+		);
+
+		nvl = fnvpair_value_nvlist(elem);
+		cval = fnvlist_lookup_string(nvl, ZPROP_VALUE);
+
+		PyObject *val = PyUnicode_FromString(cval);
+		if (val == NULL) {
+			Py_DECREF(d);
+			return NULL;
+		}
+
+		if (PyDict_SetItemString(d, name, val) < 0)  {
+			Py_DECREF(d);
+			Py_DECREF(val);
+			return NULL;
+		}
+
+		Py_DECREF(val);
+	}
+
+	return d;
+}
+
+/*
+ * Convert a python dictionary of user props into an nvlist
+ * for insertion as user properties.
+ */
+nvlist_t *py_userprops_dict_to_nvlist(PyObject *pyprops)
+{
+	nvlist_t *nvl = fnvlist_alloc();
+	PyObject *key, *value;
+	Py_ssize_t pos = 0;
+
+	while (PyDict_Next(pyprops, &pos, &key, &value)) {
```

Comment:
You probably want to check `pyprops` is a valid dictionary? (i.e. `PyDict_CheckExact` (or maybe just `PyDict_Check`)

Reply:
Ah. I moved files. This isn't new code.

Reply:
I'll fix though.